doubly_linked_list/doubly_linked_list.py

```diff
@@ -1,68 +1,161 @@
+
+
 class ListNode:
-  def __init__(self, value, prev=None, next=None):
-    self.value = value
-    self.prev = prev
-    self.next = next
+    def __init__(self, value, prev=None, next=None):
+        self.value = value
+        self.prev = prev
+        self.next = next
 
-  """Wrap the given value in a ListNode and insert it
+    """Wrap the given value in a ListNode and insert it
   after this node. Note that this node could already
   have a next node it is point to."""
-  def insert_after(self, value):
-    current_next = self.next
-    self.next = ListNode(value, self, current_next)
-    if current_next:
-      current_next.prev = self.next
 
-  """Wrap the given value in a ListNode and insert it
+    def insert_after(self, value):
+        current_next = self.next
+        self.next = ListNode(value, self, current_next)
+        if current_next:
+            current_next.prev = self.next
+
+    """Wrap the given value in a ListNode and insert it
   before this node. Note that this node could already
   have a previous node it is point to."""
-  def insert_before(self, value):
-    current_prev = self.prev
-    self.prev = ListNode(value, current_prev, self)
-    if current_prev:
-      current_prev.next = self.prev
 
-  """Rearranges this ListNode's previous and next pointers
+    def insert_before(self, value):
+        current_prev = self.prev
+        self.prev = ListNode(value, current_prev, self)
+        if current_prev:
+            current_prev.next = self.prev
+
+    """Rearranges this ListNode's previous and next pointers
   accordingly, effectively deleting this ListNode."""
-  def delete(self):
-    if self.prev:
-      self.prev.next = self.next
-    if self.next:
-      self.next.prev = self.prev
+
+    def delete(self):
+        if self.prev:
+            self.prev.next = self.next
+        if self.next:
+            self.next.prev = self.prev
+
 
 """Our doubly-linked list class. It holds references to
 the list's head and tail nodes."""
+
+
 class DoublyLinkedList:
-  def __init__(self, node=None):
-    self.head = node
-    self.tail = node
-    self.length = 1 if node is not None else 0
+    def __init__(self, node=None):
+        self.head = node
+        self.tail = node
+        self.length = 1 if node is not None else 0
+
+    def __len__(self):
+        return self.length
+
+    """
+    For debugging
+    """
+
+    def __repr__(self):
+        node = self.head
+        string = ""
+        if node is None:
+            return "[ ]"
+        while node.next is not None:
+            string += f"{node.value}, "
+            node = node.next
+        return f"[ {string}{node.value} ]"
 
-  def __len__(self):
-    return self.length
+    def add_to_head(self, value):
+        if self.length == 0:
+            node = ListNode(value)
+            self.head = node
+            self.tail = node
+        else:
+            self.head.insert_before(value)
+            self.head = self.head.prev
+        self.length += 1
 
```

Comment:
great job figuring out that insert_before does not set the value of the actual head and you still need to do that.
Do you know how to write the solution without the helper function "insert_before"?